Scope: minimize the number of gets

danielbeardsley opened the pull request:
If getting the scope prefix value is a miss, then we all keys under the scope are by definition a miss, so we don't need to bother looking.

This defers the generating and storing of the scope prefix value till we actually need it (on a set). In practice, this will reduce the number of gets if using getMultiple() or if you don't always *set()* on a miss.

danielbeardsley committed:
Scope: minimize the number of gets

If getting the scope prefix value is a miss, then we all keys under the
scope are by definition a miss, so we don't need to bother looking.

This defers the generating and storing of the scope prefix value till
we actually need it (on a set). In practice, this will reduce the number
of gets if using getMultiple() or if you don't always *set()* on a miss.

danielbeardsley committed:
Actions: update to latest version of checkout

masonmcelvain commented on tests/ScopeTest.php:
```diff
@@ -43,6 +43,46 @@ public function testScope() {
+         ->with($this->stringContains($scope))
+         ->willReturn(Matryoshka\Backend::MISS);
+      $scopedCache = new Matryoshka\Scope($mockBackend, $scope);
+
+      $key = (string)rand(1, 100000);
+      $this->assertNull($scopedCache->get($key));
+   }
+
+   public function testScopeShortCircuitGetMultiple() {
+      $scope = 'scope';
+      $mockBackend = $this->getMockBuilder(Matryoshka\Ephemeral::class)
+            ->setMethods(['get'])
+            ->getMock();
+
+      // Assert get() is called only once because a missing scope value means
+      // that all underlying values are also missing.
+      $mockBackend->expects($this->once())
+         ->method('get')
+         ->with($this->stringContains($scope))
+         ->willReturn(Matryoshka\Backend::MISS);
+      $scopedCache = new Matryoshka\Scope($mockBackend, $scope);
+
+      $key = (string)rand(1, 100000);
+      $keys = [$key => 'no matter'];
+      $this->assertSame(
+         [[$key => Matryoshka\Backend::MISS], $keys],
+         $scopedCache->getMultiple($keys)
+      );
+   }
```
This getMultiple test passes on master, so I don't think it enforces the behavior we're expecting it to.

danielbeardsley replied:
Good call... I thought it would complain if any other methods were called. I guess I should have it assert `getMutliple` isn't called under the hood.

masonmcelvain commented on library/iFixit/Matryoshka/Scope.php:
```diff
@@ -21,13 +21,16 @@ public function getPrefix() {
+         if ($scopeValue === self::MISS && !$generateOnMiss) {
+            return self::MISS;
+         }
+         if ($reset || ($scopeValue === self::MISS && $generateOnMiss)) {
```
This guard seems unnecessarily complex - if `$reset` is truthy then `$scopeValue` will be MISS at this point

```suggestion
         if ($scopeValue === self::MISS && $generateOnMiss) {
```

danielbeardsley replied:
Good call. I refactored this several times and left it in a non-optimal state.

masonmcelvain commented on library/iFixit/Matryoshka/Scope.php:
```diff
@@ -21,13 +21,16 @@ public function getPrefix() {
+         }
+         if ($reset || ($scopeValue === self::MISS && $generateOnMiss)) {
+            $scopeValue = substr(md5(microtime() . $this->scopeName), 0, 16);
+            $this->backend->set($this->getScopeKey(), $this->scopePrefix);
```
Isn't `$this->scopePrefix` still null at this point? It seems like we want to point the scopeKey at the scopeValue


```suggestion
            $this->backend->set($this->getScopeKey(), $scopeValue);
```

danielbeardsley replied:
Ooof.. This was way off. Thanks. I refactored it and added a test.

masonmcelvain reviewed (dismissed):
CR 👏🏻 but dev_block 👍🏻 on what might be a bug

masonmcelvain commented on library/iFixit/Matryoshka/Scope.php:
```diff
@@ -52,4 +55,20 @@ public function deleteScope(): bool {
+
+   public function get($key) {
+      // If the scope prefix doesn't exist, all keys in this scope are a miss.
+      if (!$this->getScopePrefix(generateOnMiss: false)) {
```
`!null` is the same as `null === null` here, I think the bot is wrong

danielbeardsley committed:
ScopeTest: assert getMultiple() isn't called

The Short circuit should prevent getMultiple from being called and we
weren't explicitly check for that.

danielbeardsley committed:
Scope: Fix a bug!

We weren't using the $scopeValue we just created when doing the set().

Also, refactor it for clarity.

Add a test to ensure we're setting the value.

masonmcelvain commented on library/iFixit/Matryoshka/Scope.php:
```diff
@@ -21,14 +21,20 @@ public function getPrefix() {
+            }
+         } else {
+            $this->scopePrefix = "{$scopeValue}-";
+         }
```
Minor: I think we could de-duplicate the `$this->scopePrefix = "{$scopeValue}-";` assignment here

```suggestion
         if ($scopeValue === self::MISS) {
            if ($generateOnMiss) {
               $scopeValue = substr(md5(microtime() . $this->scopeName), 0, 16);
               $this->backend->set($this->getScopeKey(), $scopeValue);
            } else {
               return self::MISS;
            }
         }
         $this->scopePrefix = "{$scopeValue}-";
```

danielbeardsley replied:
Fixed!

masonmcelvain reviewed (dismissed):
CR 👏🏻 Looks much better, thanks!

kwiens commented:
un_dev_block 👍 Looks like this is ready to test again.

kwiens commented:
@danielbeardsley Does this need testing, or is it ready to merge?

danielbeardsley committed:
unit tests: use native exception assertions

so phpunit stops complaining

danielbeardsley committed:
Scope: DRY up the assignment

danielbeardsley commented:
It passes tests. So, I guess QA :+1:

masonmcelvain reviewed (approved):
CR 👏🏻
